### PR TITLE
Support Aave v3 aToken balance overrides

### PR DESCRIPTION
**Pull request by squadgazzz**
# Description
Quote verification was silently failing for any Aave v3 aToken as the sell token (e.g. aEthWETH → WETH) because the balance-override mechanism assumes the value at the balance storage slot is what `balanceOf` returns. Aave v3 aTokens break both assumptions: `balanceOf` applies a `rayMul(scaled, liquidityIndex)` scaling, and storage is packed `UserState { uint128 balance; uint128 additionalData }`. The auto-detector's verify step therefore never matched, returned `NotFound`, and the trade verifier silently skipped the override, producing the production revert `execution reverted: trader does not have enough sell token` visible in barn logs for aEthWETH quotes today (reproduced on Tenderly: https://dashboard.tenderly.co/cow-protocol/barn/simulator/1e69fa91-9496-44d1-9aec-a0e34166f9df).

# Changes
- New `Strategy::AaveV3AToken { target_contract, pool, underlying, map_slot }` variant in `configs::balance_overrides::Strategy` and a corresponding async resolver on `BalanceOverrides` that fetches the current `getReserveNormalizedIncome` from the Aave v3 Pool, rayDivs the requested amount (round-half-up, bit-for-bit compatible with `WadRayMath.rayDiv`), and writes it into the low 128 bits of the packed `_userState` slot.
- Auto-detector is now Aave-aware: the detector probes the token with `UNDERLYING_ASSET_ADDRESS()` + `POOL()` and then calls `pool.getReserveData(underlying)` and verifies the returned `aTokenAddress` equals the probed token — an identity check that only passes for tokens the pool itself has registered as an aToken for their underlying, preventing rogue contracts that merely implement the aToken selectors from being accepted. When the probe succeeds, mapping-style slots are also offered as `AaveV3AToken` candidates and verified with a 1-wei round-trip tolerance instead of strict equality. No hardcoded per-token list is needed. aEthWETH and every other Aave v3 aToken on every chain (mainnet, Arbitrum, Base, Gnosis, Polygon, BNB, Linea, Plasma, Ink, Avalanche, etc.) are picked up automatically the first time they're quoted.
- Shared Aave math lives in a new `balance-overrides::aave` module so the production override builder and the detector probe/verify use identical arithmetic.
- `trade_verifier::prepare_state_overrides` now emits `tracing::warn!` when the spardose balance override for the sell token can't be resolved — mirroring the existing warn on the buy-side path so future missing overrides surface in logs instead of only showing up as downstream reverts.

# Aave v4 note
Aave v4's user-facing deposit receipt is `TokenizationSpoke`, an ERC-4626 vault built on OpenZeppelin's `ERC20Upgradeable` — not a scaled aToken. `balanceOf` is the plain OZ `_balances[user]` mapping (no `rayMul` scaling), and the contract doesn't expose `UNDERLYING_ASSET_ADDRESS()` or `POOL()`. Our probe therefore correctly rejects v4 Spokes and the detector falls through to the existing `SolidityMapping` / `DirectSlot` heuristics, which handle them as standard ERC-20s. No v4-specific work is needed here.
Source: <https://github.com/aave/aave-v4/blob/main/src/spoke/TokenizationSpoke.sol>

# How to test
Unit tests:
- `balance-overrides::tests::a_token_balance_override_bug_reproduction` — pins the arithmetic: writing the raw amount (what the old strategies do) makes `balanceOf` return `rayMul(amount, index)` (off by ~6e16 wei at aEthWETH's current index); writing `rayDiv(amount, index)` round-trips within 1 wei.
- `balance-overrides::aave::tests` — mock-provider tests for the probe: accepts a valid aToken, rejects when either selector reverts, rejects when the claimed pool doesn't look like Aave v3, rejects when the pool registers a *different* aToken for the declared underlying.
- `balance-overrides::tests::aave_v3_a_token_override_scales_amount_and_writes_low_128` — mock-provider integration test for the override builder.

e2e local + forked node tests:
- `balance-overrides::detector::tests::detects_aave_v3_a_token_mainnet` — asserts the detector returns `AaveV3AToken { pool=0x87870bca…4fa4e2, underlying=WETH, map_slot=52 }` for aEthWETH.
- `balance-overrides::tests::aave_v3_a_token_override_mainnet_roundtrip` — applies the override via `eth_call` against real aEthWETH and asserts `balanceOf(holder) ≈ amount`.
- `e2e::quote_verification::forked_node_mainnet_aave_atoken_quote` — same round-trip against an anvil mainnet fork, exercising the full `BalanceOverrides::state_override` path.

# Follow-up items
- **Registry-based detection.** The current probe costs 3 `eth_call`s per cold-cache token. An alternative is to enumerate all reserves once per chain from Aave's `AaveProtocolDataProvider` (or equivalent), cache the full list of aTokens, and do a pure `HashMap` lookup on each quote. That would make detection amortised O(1) and provide the strongest identity guarantee (taken from Aave's own registry), at the cost of introducing a per-chain constant for the DataProvider address, a refresh cadence for new markets, and handling multiple Aave v3 markets per chain (e.g. mainnet has Ethereum + Prime). Probably worth doing once aToken quote volume warrants the optimisation.
- **Cache the accrued liquidity index for ~1 block.** Each aToken quote currently re-fetches `getReserveNormalizedIncome`. The index drifts slowly (fractions of a wei per block), so caching for 6–12 s would drop the per-quote cost to zero without meaningful accuracy loss.
- **Aave v2 support.** Same scaling + packed-storage shape; different `LendingPool` interface. v2 is deprecated but still has markets; worth revisiting if any start to attract volume.

### CODE REVIEW

File: crates/balance-overrides/src/aave.rs [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+    ///
+    /// Source: <https://github.com/aave-dao/aave-v3-origin/blob/main/src/contracts/interfaces/IAToken.sol>
+    interface IAaveV3AToken {
+        function UNDERLYING_ASSET_ADDRESS() external view returns (address);
+        function POOL() external view returns (address);
+    }
+}
+
+/// Ray (1e27) — Aave's 27-decimal fixed-point unit.
+pub const RAY: U256 = U256::from_limbs([0x9fd0803ce8000000, 0x33b2e3c, 0, 0]);
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/aave.rs:76]: Would be good to provide links wherever possible.

File: crates/balance-overrides/src/aave.rs [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+    let packed: U256 = ((additional_data & mask) << 128) | (balance & mask);
+    B256::new(packed.to_be_bytes::<32>())
+}
+
+/// Probes whether `token` is an Aave v3 aToken. The token self-declares its
+/// `UNDERLYING_ASSET_ADDRESS()` and `POOL()`; we then ask the pool for the
+/// registered `aTokenAddress` of that underlying and require it to equal the
+/// probed token. This catches both "contract doesn't look anything like an
+/// aToken" and "contract impersonates the aToken interface but is not
+/// registered with the pool for its declared underlying."
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/aave.rs]: This reads very AI generated. 😅

File: crates/balance-overrides/src/aave.rs [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+        state_diff: Some(iter::once((slot, value)).collect()),
+        ..Default::default()
+    };
+    Some((a_token, state_override))
+}
+
+/// Helper: `eth_call` a zero-argument view returning `address`. Returns
+/// `None` on any RPC error or decode failure (including the common case of
+/// the token not exposing this selector at all).
+async fn call_address(web3: &Web3, to: Address, calldata: Vec<u8>) -> Option<Address> {
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/aave.rs]: This function looks very awkward. Wouldn't it be nicer to do the call through the ABI structs (e.g. `IAaveV3AToken::POOLCall`)?
That way alloy automatically deserializes the result for you correctly.

**Comment by squadgazzz** [comment]
@claude review

File: crates/balance-overrides/src/detector.rs [hunk 4]
```diff
@@ -92,6 +93,20 @@ fn create_strategies_from_slots(
...
+            // the probe was a false positive (another contract happens to
+            // expose the same two selectors).
+            if let Some(info) = aave_info.as_ref() {
+                heuristic_strategies.push(Strategy::AaveV3AToken {
+                    target_contract: *contract,
+                    pool: info.pool,
+                    underlying: info.underlying,
+                    map_slot: U256::from(map_slot_index),
+                });
+            }
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/detector.rs]: With a strategy specialized for aave tokens we already know which storage slot the address would be, right? In that case we only have to probe 1 storage slot and save a bunch of RPC requests.

File: crates/e2e/tests/e2e/quote_verification.rs [hunk 5]
```diff
@@ -790,3 +812,75 @@ async fn trace_based_balance_detection(web3: Web3) {
...
+    let mut hardcoded = HashMap::new();
+    hardcoded.insert(
+        a_eth_weth,
+        Strategy::AaveV3AToken {
+            target_contract: a_eth_weth,
+            pool: aave_v3_pool,
+            underlying: weth,
+            map_slot: U256::from(52),
+        },
+    );
```
Comment (jmg-duarte) [inline crates/e2e/tests/e2e/quote_verification.rs]: nit - removes the mut
```suggestion
    let hardcoded = HashMap::from_iter(&[(
    	a_eth_weth,
        Strategy::AaveV3AToken {
            target_contract: a_eth_weth,
            pool: aave_v3_pool,
            underlying: weth,
            map_slot: U256::from(52),
        })
    ]);
```

File: crates/e2e/tests/e2e/quote_verification.rs [hunk 6]
```diff
@@ -790,3 +812,75 @@ async fn trace_based_balance_detection(web3: Web3) {
...
+/// reported balance matches the requested amount within one wei of ray
+/// rounding. This is the property `TradeVerifier` relies on when using
+/// the override to fund the spardose.
+async fn aave_atoken_quote_verification(web3: Web3) {
+    use {
+        ::alloy::rpc::types::state::StateOverride,
+        balance_overrides::BalanceOverrideRequest,
+        contracts::ERC20,
+        std::collections::HashMap,
+    };
```
Comment (jmg-duarte) [inline crates/e2e/tests/e2e/quote_verification.rs]: does it really need to be inside the function?

File: crates/balance-overrides/src/detector.rs [hunk 7]
```diff
@@ -324,8 +368,32 @@ impl Detector {
         // Use a unique test value to verify this strategy works
         let test_balance = U256::from(0x1337_1337_1337_1337_u64);
 
-        // Create state override using the strategy
-        let overrides = strategy.state_override(&holder, &test_balance);
+        // Create state override using the strategy. `AaveV3AToken` needs an
+        // async call to the pool to compute the scaled storage value, so we
+        // reuse the production builder rather than `StrategyExt::state_override`
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/detector.rs]: The module comment on `aave.rs` also references "production". What is that supposed to mean in this context?

File: crates/e2e/tests/e2e/quote_verification.rs [hunk 8]
```diff
@@ -790,3 +812,75 @@ async fn trace_based_balance_detection(web3: Web3) {
...
+        .state(overrides)
+        .call()
+        .await
+        .unwrap();
+
+    let diff = if reported > amount {
+        reported - amount
+    } else {
+        amount - reported
+    };
```
Comment (jmg-duarte) [inline crates/e2e/tests/e2e/quote_verification.rs]: Wouldn't (reported - amount).abs() work? or even the abs_diff like you used earlier

File: crates/balance-overrides/src/lib.rs [hunk 9]
```diff
@@ -1,11 +1,16 @@
+pub mod aave;
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/lib.rs]:
```suggestion
mod aave;
```

File: crates/e2e/tests/e2e/quote_verification.rs [hunk 10]
```diff
@@ -790,3 +812,75 @@ async fn trace_based_balance_detection(web3: Web3) {
...
+            map_slot: U256::from(52),
+        },
+    );
+    let balance_overrides = BalanceOverrides {
+        hardcoded,
+        detector: None,
+        web3: Some(web3.clone()),
+    };
+
+    let amount = U256::from(5_000_000_000_000_000_000u128); // 5 aEthWETH
```
Comment (jmg-duarte) [inline crates/e2e/tests/e2e/quote_verification.rs]:
```suggestion
    let amount = 5u64.eth(); // 5 aEthWETH
```
with the respective trait

File: crates/balance-overrides/src/lib.rs [hunk 11]
```diff
@@ -75,6 +78,12 @@ impl StrategyExt for Strategy {
                 target_contract,
                 slot,
             } => (target_contract, *slot),
+            // AaveV3AToken requires an async call to fetch the current
+            // normalized income, so it is handled separately in
+            // `BalanceOverrides::state_override`.
+            Self::AaveV3AToken { .. } => unreachable!(
+                "AaveV3AToken strategy must be resolved asynchronously, not via StrategyExt"
+            ),
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/lib.rs]: This seems fragile. I didn't check but I assume there must be short circuiting logic somewhere so that we never trigger this.
I suppose it would be safer to just make this trait async and keep the logic here.

File: crates/balance-overrides/src/detector.rs [hunk 12]
```diff
@@ -433,6 +507,32 @@ mod tests {
         );
     }
 
+    /// The detector recognises aEthWETH as an Aave v3 aToken (via
+    /// `UNDERLYING_ASSET_ADDRESS()` + `POOL()` probes) and returns the
+    /// right `AaveV3AToken` strategy, scaling and all, without any
+    /// hardcoded per-token config.
+    /// Set `NODE_URL` environment to a mainnet RPC URL.
```
Comment (jmg-duarte) [inline crates/balance-overrides/src/detector.rs]: should we make this a forked test instead?
it would make this test run more often

File: crates/balance-overrides/src/lib.rs [hunk 13]
```diff
@@ -104,6 +113,9 @@ pub struct BalanceOverrides {
     /// The balance override detector and its cache. Set to `None` if
     /// auto-detection is not enabled.
     pub detector: Option<(Detector, DetectorCache)>,
+    /// Optional web3 handle used by strategies that need runtime on-chain
+    /// reads (e.g. `AaveV3AToken` fetching the reserve's normalized income).
+    pub web3: Option<Web3>,
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/lib.rs]: I'm a bit surprised that we need to introduce the `Web3` here. All other strategies also need a `Web3` or else they couldn't do the underlying RPC requests. 🤔

File: crates/balance-overrides/src/detector.rs [hunk 14]
```diff
@@ -340,12 +408,18 @@ impl Detector {
...
+        // aToken balances round-trip through rayDiv/rayMul so the reported
+        // balance may differ from the written value by one wei. Every other
+        // strategy must match exactly.
+        let balance_matches = if matches!(strategy, Strategy::AaveV3AToken { .. }) {
+            balance.abs_diff(test_balance) <= U256::from(1u64)
         } else {
-            Err(DetectionError::NotFound)
-        }
+            balance == test_balance
+        };
```
Comment (jmg-duarte) [inline crates/balance-overrides/src/detector.rs]: nit: use a match statement instead, feels more native to what we're used in this case imo

File: crates/balance-overrides/src/detector.rs [hunk 15]
```diff
@@ -324,8 +368,32 @@ impl Detector {
...
+                    *map_slot,
+                    holder,
+                    test_balance,
+                )
+                .await
+                .ok_or(DetectionError::NotFound)?;
+                std::iter::once((target, account_override)).collect()
+            }
+            _ => strategy.state_override(&holder, &test_balance),
+        };
```
Comment (jmg-duarte) [inline crates/balance-overrides/src/detector.rs]: IMO this could be extracted into a separate function

File: crates/balance-overrides/src/aave.rs [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+    B256::new(packed.to_be_bytes::<32>())
+}
+
+/// Probes whether `token` is an Aave v3 aToken. The token self-declares its
+/// `UNDERLYING_ASSET_ADDRESS()` and `POOL()`; we then ask the pool for the
+/// registered `aTokenAddress` of that underlying and require it to equal the
+/// probed token. This catches both "contract doesn't look anything like an
+/// aToken" and "contract impersonates the aToken interface but is not
+/// registered with the pool for its declared underlying."
+pub async fn probe_a_token(web3: &Web3, token: Address) -> Option<(Address, Address)> {
```
Comment (jmg-duarte) [inline crates/balance-overrides/src/aave.rs]: I know this is meant to be used as `aave::prove_a_token` but the name is nonetheless, confusing
```suggestion
pub async fn probe_aave_token(web3: &Web3, token: Address) -> Option<(Address, Address)> {
```

File: crates/balance-overrides/src/aave.rs [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+    alloy_eips::BlockId,
+    alloy_primitives::{Address, B256, Bytes, TxKind, U256, keccak256},
+    alloy_provider::Provider,
+    alloy_rpc_types::{TransactionInput, TransactionRequest, state::AccountOverride},
+    alloy_sol_types::{SolCall, sol},
+    ethrpc::Web3,
+    std::iter,
+};
+
+sol! {
```
Comment (jmg-duarte) [inline crates/balance-overrides/src/aave.rs:18]: moving this into an abi json or whatever it needs would allow us to introspect if needed + still keeps the macros for contracts at bay

File: crates/balance-overrides/src/aave.rs [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+            id: 0,
+            aTokenAddress: a_token,
+            stableDebtTokenAddress: Address::ZERO,
+            variableDebtTokenAddress: Address::ZERO,
+            interestRateStrategyAddress: Address::ZERO,
+            accruedToTreasury: 0,
+            unbacked: 0,
+            isolationModeTotalDebt: 0,
+        };
+        format!("0x{}", alloy_primitives::hex::encode(data.abi_encode()))
```
Comment (jmg-duarte) [inline crates/balance-overrides/src/aave.rs]: use encode_prefixed instead

File: crates/balance-overrides/src/aave.rs [hunk 19]
```diff
@@ -0,0 +1,359 @@
...
+mod tests {
+    use {
+        super::*,
+        alloy_primitives::address,
+        alloy_provider::mock::Asserter,
+        alloy_sol_types::SolValue,
+    };
+
+    fn encode_address(addr: Address) -> String {
+        format!("0x{:0>64x}", U256::from_be_bytes(addr.into_word().0))
```
Comment (jmg-duarte) [inline crates/balance-overrides/src/aave.rs]: doesn't this work?
```suggestion
        hex::encode_prefixed(U256::from_be_bytes(addr.into_word().0))
```

File: crates/balance-overrides/src/aave.rs [hunk 20]
```diff
@@ -0,0 +1,359 @@
...
+        let underlying = address!("c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2");
+
+        let asserter = Asserter::new();
+        // 1. UNDERLYING_ASSET_ADDRESS() → underlying
+        asserter.push_success(&encode_address(underlying));
+        // 2. POOL() → pool
+        asserter.push_success(&encode_address(pool));
+        // 3. pool.getReserveData(underlying) → ReserveData { aTokenAddress: a_token, ..
+        //    }
+        asserter.push_success(&encode_reserve_data(a_token));
```
Comment (jmg-duarte) [inline crates/balance-overrides/src/aave.rs]: Since this is the general test pattern here, I think you could cover the "walk" at the test module root (so around line 247) and that would cover for the others too

**Review by jmg-duarte — COMMENTED** [review]
logic wise, everything looks sound to me

File: crates/balance-overrides/src/aave.rs (hunk 17 above)
Comment (squadgazzz) [reply crates/balance-overrides/src/aave.rs:18]: Keeping the inline `sol!` for this PR. The interface surface is small (2 functions + 1 struct), used only here, and we don't need deployment bytecode — vendoring + generated bindings would add more overhead than it saves.

If we end up doing registry-based detection (less likely), we'd want `IPoolAddressesProvider` + `AaveProtocolDataProvider` + full `IPool`, which is the right point to vendor a proper set. Source links in the `sol!` block point at the canonical contracts in the meantime.

File: crates/balance-overrides/src/aave.rs [hunk 21]
```diff
@@ -0,0 +1,338 @@
...
+/// underlying)` pair. Accepts the token iff the pool registers it as the
+/// aToken for its declared underlying — rogue contracts implementing the
+/// aToken selectors aren't enough.
+pub async fn probe_aave_token(web3: &Web3, token: Address) -> Option<(Address, Address)> {
+    let a_token = IAaveV3AToken::new(token, web3.provider.clone());
+    let underlying_call = a_token.UNDERLYING_ASSET_ADDRESS();
+    let pool_call = a_token.POOL();
+    let (underlying, pool) = tokio::join!(underlying_call.call(), pool_call.call());
+    let underlying = underlying.ok()?;
+    let pool = pool.ok()?;
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/aave.rs]:
```suggestion
    let (underlying, pool) = tokio::try_join!(underlying_call.call(), pool_call.call()).ok()?;
```

File: crates/balance-overrides/src/aave.rs [hunk 22]
```diff
@@ -0,0 +1,338 @@
...
+    let reserve = fetch_reserve_data(web3, pool, underlying).await?;
+    (reserve.aTokenAddress == token).then_some((pool, underlying))
+}
+
+/// Fetches `getReserveData(underlying)` from an Aave v3 Pool.
+async fn fetch_reserve_data(
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/aave.rs]: nit: feels like this might as well be inlined into the only function that calls it.

File: crates/balance-overrides/src/aave.rs [hunk 23]
```diff
@@ -0,0 +1,338 @@
...
+) -> Option<ReserveData> {
+    IAaveV3Pool::new(pool, web3.provider.clone())
+        .getReserveData(underlying)
+        .call()
+        .await
+        .ok()
+}
+
+/// Fetches `getReserveNormalizedIncome(underlying)` from an Aave v3 Pool.
+pub async fn fetch_normalized_income(
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/aave.rs]: same here. The function called on the contract struct already gives you all the context you need, no?

File: crates/balance-overrides/src/aave.rs [hunk 24]
```diff
@@ -0,0 +1,338 @@
...
+pub async fn build_override(
+    web3: &Web3,
+    a_token: Address,
+    pool: Address,
+    underlying: Address,
+    map_slot: U256,
+    holder: Address,
+    amount: U256,
+) -> Option<(Address, AccountOverride)> {
+    let index = match fetch_normalized_income(web3, pool, underlying).await {
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/aave.rs]: nit: this and the next `match` could use a `let-else`

File: crates/balance-overrides/src/aave.rs [hunk 25]
```diff
@@ -0,0 +1,338 @@
...
+            tracing::warn!(
+                ?a_token,
+                %amount,
+                %index,
+                "ray_div overflow computing AaveV3AToken scaled balance"
+            );
+            return None;
+        }
+    };
+    let slot = mapping_slot_hash(&holder, &map_slot.to_be_bytes::<32>());
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/aave.rs]: nit: no need to specify this
```suggestion
    let slot = mapping_slot_hash(&holder, &map_slot.to_be_bytes());
```

File: crates/balance-overrides/src/aave.rs [hunk 26]
```diff
@@ -0,0 +1,338 @@
...
+    use {
+        super::*,
+        alloy_primitives::{address, hex},
+        alloy_provider::mock::Asserter,
+        alloy_sol_types::SolValue,
+    };
+
+    fn encode_address(addr: Address) -> String {
+        hex::encode_prefixed(U256::from_be_bytes(addr.into_word().0).to_be_bytes::<32>())
+    }
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/aave.rs:203]: This could be replaced entirely with `Address::encode_hex_with_prefix()`, no?

**Review by jmg-duarte — APPROVED** [review]

File: crates/balance-overrides/src/aave.rs [hunk 27]
```diff
@@ -0,0 +1,338 @@
...
+
+    /// Encodes a `ReserveData` struct as the ABI return payload the pool
+    /// would produce. All fields other than `aTokenAddress` are zero — we
+    /// only care about that one for the probe.
+    fn encode_reserve_data(a_token: Address) -> String {
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/aave.rs:208]: Instead of building all this mocking machinery wouldn't it be a lot simpler to just to a forked test on an actual token?

This actually comes down to the goal of the e2e tests: Test at the highest possible level of abstraction to avoid breaking 10 different tests when you change some interface in something that should be an implementation detail.

File: crates/balance-overrides/src/detector.rs [hunk 28]
```diff
@@ -124,17 +128,55 @@ impl Detector {
...
+                map_slot: U256::from(aave::USER_STATE_SLOT),
+            };
+            if self
+                .verify_strategy(token, holder, &candidate)
+                .await
+                .is_ok()
+            {
+                tracing::debug!(
+                    ?token,
+                    "detected Aave v3 aToken via fast-path (no debug_traceCall)"
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/detector.rs]: Please do a sanity check before opening a PR with AI generated code. Why would we have to log here that we did not do a `debug_traceCall`?

File: crates/balance-overrides/src/detector.rs [hunk 29]
```diff
@@ -124,17 +128,55 @@ impl Detector {
...
+        // moved `_userState` to a different slot won't verify here and
+        // will fall through to the generic trace-based path, which only
+        // ever returns non-Aave strategies; such a fork needs an explicit
+        // hardcoded config entry.
+        if let Some((pool, underlying)) = aave::probe_aave_token(&self.web3, token).await {
+            let candidate = Strategy::AaveV3AToken {
+                target_contract: token,
+                pool,
+                underlying,
+                map_slot: U256::from(aave::USER_STATE_SLOT),
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/detector.rs]: If we always pass in a constant value can't we just have the function that takes the `Strategy::AaveV3AToken` inject that when it's needed?

File: crates/balance-overrides/src/detector.rs [hunk 30]
```diff
@@ -340,12 +391,20 @@ impl Detector {
...
+    }
+}
+
+/// Is the `balance` returned by `balanceOf` after applying the override
+/// consistent with the `test_balance` we wrote? For `AaveV3AToken` we tolerate
+/// 1 wei of difference because Aave's ray-div / ray-mul round-trip is not
+/// identity by construction; every other strategy must match exactly.
+fn verified_balance_matches(strategy: &Strategy, balance: U256, test_balance: U256) -> bool {
+    match strategy {
+        Strategy::AaveV3AToken { .. } => balance.abs_diff(test_balance) <= U256::from(1u64),
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/detector.rs]:
```suggestion
        Strategy::AaveV3AToken { .. } => balance.abs_diff(test_balance) <= U256::ONE,
```

File: crates/balance-overrides/src/lib.rs [hunk 31]
```diff
@@ -45,48 +50,92 @@ pub struct BalanceOverrideRequest {
...
 }
 
+#[async_trait::async_trait]
 trait StrategyExt {
     /// Computes the storage slot and value to override for a particular token
-    /// holder and amount.
-    fn state_override(&self, holder: &Address, amount: &U256) -> AddressMap<AccountOverride>;
+    /// holder and amount. `web3` is only consulted by strategies that need
+    /// on-chain reads at override time (currently `AaveV3AToken`); other
+    /// variants ignore it and complete synchronously.
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/lib.rs:58]: If `web3` is only used by strategies could we not put the `Web3` struct into the `Strategy` itself so that it has everything it needs?

File: crates/balance-overrides/src/lib.rs [hunk 32]
```diff
@@ -448,4 +519,255 @@ mod tests {
...
     }
+
+    /// Round-half-up `rayMul` — the same formula aToken's `balanceOf` applies
+    /// to convert scaled storage into the reported display balance. Only used
+    /// by the bug-reproduction test below.
+    fn ray_mul(a: U256, b: U256) -> U256 {
+        (a * b + (RAY >> 1)) / RAY
+    }
+
+    /// Reproduction of the aEthWETH bug in pure math: without the
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/lib.rs]: These tests can probably also just be replaced with 1 higher level test that asserts that the balance can be correctly overriden for a given aave token.

File: crates/configs/src/balance_overrides.rs [hunk 33]
```diff
@@ -36,6 +36,20 @@ pub enum Strategy {
...
         slot: B256,
     },
+    /// Balance override strategy for Aave v3 aTokens whose `balanceOf` applies
+    /// a scaling factor (the reserve's normalized income) and whose storage
+    /// packs `UserState { uint128 balance; uint128 additionalData }` into a
+    /// single slot. The override is resolved at call time by fetching the
+    /// current normalized income from the Aave pool and dividing the target
+    /// amount by it (ray division), so that `balanceOf` returns the desired
+    /// amount. The high 128 bits of the slot (`additionalData`) are zeroed,
+    /// which is safe for fresh holders like the spardose.
```
Comment (MartinquaXD) [inline crates/configs/src/balance_overrides.rs:46]: Could also use a link.

File: crates/balance-overrides/src/aave.rs (hunk 26 above)
Comment (jmg-duarte) [reply crates/balance-overrides/src/aave.rs:203]: i dont think so, the difference is in the number of bytes
the u256 will pad the remaining 12 with zeros, the address will return just return 20 bytes
so if the expectation is 32bytes with an address inside, this is the way to go

Comment (squadgazzz) [reply crates/balance-overrides/src/aave.rs:203]: Final form is `hex::encode_prefixed(addr.into_word())`, which goes through `B256` (32 bytes: 12-byte zero pad + 20 address bytes), matching the ABI-encoded address return that the Asserter feeds to alloy's decoder. Agreed with @jmg-duarte's reasoning; `Address::encode_hex_with_prefix()` would produce 20 bytes and break decoding.

File: crates/balance-overrides/src/aave.rs (hunk 27 above)
Comment (squadgazzz) [reply crates/balance-overrides/src/aave.rs:208]: Partially addressed. The happy-path probe mock was dropped (`probe_aave_token_accepts_valid_atoken`), and the forked-e2e `forked_node_mainnet_aave_atoken_detection` now covers that plus the end-to-end override application.

Kept three rejection-path mock tests:

* `probe_aave_token_rejects_when_underlying_call_reverts`
* `probe_aave_token_rejects_when_pool_is_not_aave`
* `probe_aave_token_rejects_when_pool_registers_a_different_atoken`

The first two reject branches can't be reached against real mainnet state without deploying a scaffold contract to the fork, and the third (rogue contract impersonating an aToken for an underlying the pool has registered elsewhere) likewise has no real-world candidate on any chain we fork. The probe_asserter helper + the two ABI encoders are ~30 lines total, and the tests run in sub-millisecond, so they stay.

Happy to drop the last three too if you'd prefer a strict 'forked e2e only' policy on this module.

File: crates/balance-overrides/src/lib.rs (hunk 31 above)
Comment (squadgazzz) [reply crates/balance-overrides/src/lib.rs:58]: Considered this and the tension is that `Strategy` lives in `configs::balance_overrides` as a `serde`-deserialized config type (loaded from TOML). A `Web3` handle cannot sit inside a serde type, so carrying it on the enum would require splitting into a config variant and a runtime variant with an attachment pass on `BalanceOverrides` construction.

That's a fair chunk of plumbing for a one-off `Option<&Web3>` parameter on the trait. Leaving it as the extra arg for now.

File: crates/balance-overrides/src/lib.rs (hunk 32 above)
Comment (squadgazzz) [reply crates/balance-overrides/src/lib.rs]: Partially addressed (full address pending push, the commit is local):

* Dropped `a_token_balance_override_bug_reproduction` and its `ray_mul` helper.
* Dropped the `#[ignore]` `aave_v3_a_token_override_mainnet_roundtrip` (was manually-run only, covered by forked-e2e detection + quote tests).

Kept:

* `ray_div_edge_cases` (pure-math edge cases for the helper; 0/x and x/0 shouldn't need a forked test).
* `aave_v3_a_token_override_scales_amount_and_writes_low_128` (mock-provider integration, CI-friendly).
* `aave_v3_a_token_override_none_without_web3` (tests the `Option<&Web3>` early-out path).

If you'd also like the last two gone and pure-math-only.

File: crates/balance-overrides/src/detector.rs (hunk 29 above)
Comment (squadgazzz) [reply crates/balance-overrides/src/detector.rs]: Done. `map_slot` is gone from `Strategy::AaveV3AToken` entirely (both the `configs` enum variant and every struct literal), and `aave::build_override` injects `USER_STATE_SLOT` internally. The detector fast-path no longer sets it, and the config TOML schema no longer accepts it. Change is in a local commit that hasn't been pushed yet; will land shortly.

File: crates/balance-overrides/src/lib.rs (hunk 31 above)
Comment (MartinquaXD) [reply crates/balance-overrides/src/lib.rs:58]: I see. Given that we don't use the hardcoded configurations in practice and that our detection capabilities should now automatically find anything the hardcoded list can cover what do you think about dropping the hardcoded configuration altogether?
That way we can drop the `Deserialize` on the strategy and have the cleaner implementation with the provider inside the `Strategy`.

Given that this change would be not super small it would probably make the most sense to merge this PR as is and follow up with another clean up PR.

File: crates/balance-overrides/src/aave.rs [hunk 34]
```diff
@@ -0,0 +1,301 @@
...
+        }
+        asserter
+    }
+
+    // The happy path (probe accepts a valid aToken) is covered by the
+    // forked-e2e `forked_node_mainnet_aave_atoken_detection` test, which runs
+    // against a real mainnet fork. The tests below cover the rejection
+    // branches, which would require deploying scaffold contracts on a fork
+    // to reproduce — cheaper to simulate with the mock `Asserter`.
+
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/aave.rs]: Previously there was a test here. Now we have a comment explaining why we don't have a test here. 😓 
```suggestion
```

File: crates/balance-overrides/src/lib.rs [hunk 35]
```diff
@@ -448,4 +517,139 @@ mod tests {
...
+
+    #[test]
+    fn ray_div_edge_cases() {
+        // End-to-end numerical correctness is covered by the forked-e2e
+        // `forked_node_mainnet_aave_atoken_detection` /
+        // `forked_node_mainnet_aave_atoken_quote` tests, which apply the
+        // override against a real aEthWETH and assert `balanceOf`
+        // round-trips. Here we only pin down `ray_div` itself at the two
+        // edge cases.
+
```
Comment (MartinquaXD) [inline crates/balance-overrides/src/lib.rs]: same here
```suggestion
```

**Review by MartinquaXD — APPROVED** [review]
LGTM besides the 2 stray comments
